substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSections.java

```diff
@@ -0,0 +1,2344 @@
+        // the primary range should already have been seen
+        // and associated with a primary class entry
+        assert classEntry.primaryIndexFor(primaryRange) != null;
+        classEntry.addSubRange(subrange, subrangeEntry);
+    }
+
+    private DirEntry ensureDirEntry(String file) {
+        int pathLength = file.lastIndexOf('/');
+        if (pathLength < 0) {
+            // no path/package means use dir entry 0
+            return null;
+        }
+        String filePath = file.substring(0, pathLength);
+        DirEntry dirEntry = dirsIndex.get(filePath);
+        if (dirEntry == null) {
+            dirEntry = new DirEntry(stringTable.uniqueString(filePath));
+            dirsIndex.put(filePath, dirEntry);
+            dirs.add(dirEntry);
+        }
+        return dirEntry;
+    }
+
+    /**
+     * class from which all DWARF debug section
+     * inherit providing common behaviours.
+     */
+    // shared implementation methods to manage content creation
+    public abstract class DwarfSectionImpl extends BasicProgbitsSectionImpl {
+        public boolean debug = false;
+        public long debugTextBase = 0;
+        public long debugAddress = 0;
+        public int debugBase = 0;
+
+        public DwarfSectionImpl() {
+        }
+
+        /**
+         * creates the target byte[] array used to define the section
+         * contents.
+         *
+         * the main task of this method is to precompute the
+         * size of the debug section. given the complexity of the
+         * data layouts that invariably requires performing a dummy
+         * write of the contents, inserting bytes into a small,
+         * scratch buffer only when absolutely necessary. subclasses
+         * may also cache some information for use when writing the
+         * contents.
+         */
+        public abstract void createContent();
+
+        /**
+         * populates the byte[] array used to contain the section
+         * contents.
+         *
+         * in most cases this task reruns the operations performed
+         * under createContent but this time actually writing data
+         * to the target byte[].
+         */
+        public abstract void writeContent();
+
+        @Override
+        public boolean isLoadable() {
+            // even though we're a progbits section impl we're not actually loadable
+            return false;
+        }
+
+        public void checkDebug(int pos) {
+            // if the env var relevant to this element
+            // type is set then switch on debugging
+            String name = getSectionName();
+            String envVarName = "DWARF_" + name.substring(1).toUpperCase();
+            if (System.getenv(envVarName) != null) {
+                debug = true;
+                debugBase = pos;
+                debugAddress = debugTextBase;
+            }
+        }
+
+        protected void debug(String format, Object... args) {
+            if (debug) {
```

Comment:
We have org.graalvm.compiler.debug.DebugContext for logging. See e.g. `com.oracle.svm.hosted.ResourcesFeature#scanDirectory`